src/webhooks/registry.ts

```diff
@@ -367,105 +429,52 @@ export function createProcess(config: ConfigInterface) {
-            topic: graphqlTopic,
-          });
-
-          if (webhookEntry) {
-            try {
-              webhookEntry.webhookHandler(
-                graphqlTopic,
-                domain as string,
-                rawBody,
-              );
-              statusCode = StatusCode.Ok;
-            } catch (error) {
-              statusCode = StatusCode.InternalServerError;
-              responseError = new ShopifyErrors.InvalidWebhookError({
-                message: error.message,
-                code: StatusCode.InternalServerError,
-                statusText: 'Not Found',
-                cause: error,
-              });
-            }
-          } else {
-            statusCode = StatusCode.NotFound;
-            responseError = new ShopifyErrors.InvalidWebhookError({
-              message: `No webhook is registered for topic ${topic}`,
-              code: StatusCode.NotFound,
-              statusText: 'Not Found',
-            });
+    const webhookCheck = checkWebhookRequest(rawBody, request.headers);
+    const {webhookOk, hmac, topic, domain} = webhookCheck;
+    let {errorMessage} = webhookCheck;
+
+    if (webhookOk) {
+      const generatedHash = createHmac('sha256', config.apiSecretKey)
+        .update(rawBody, 'utf8')
+        .digest('base64');
+
+      if (safeCompare(generatedHash, hmac)) {
+        const graphqlTopic = topic.toUpperCase().replace(/\//g, '_');
+        const webhookEntry = getHandler({
+          topic: graphqlTopic,
+        });
+
+        if (webhookEntry) {
+          try {
+            webhookEntry.webhookHandler(graphqlTopic, domain, rawBody);
+            response.statusCode = StatusCode.Ok;
+          } catch (error) {
+            response.statusCode = StatusCode.InternalServerError;
+            errorMessage = error.message;
           }
         } else {
-          statusCode = StatusCode.Unauthorized;
-          responseError = new ShopifyErrors.InvalidWebhookError({
-            message: `Could not validate request for topic ${topic}`,
-            code: StatusCode.Unauthorized,
-            statusText: 'Unauthorized',
-          });
+          response.statusCode = StatusCode.NotFound;
+          errorMessage = `No webhook is registered for topic ${topic}`;
         }
+      } else {
+        response.statusCode = StatusCode.Unauthorized;
+        errorMessage = `Could not validate request for topic ${topic}`;
+      }
+    } else {
+      response.statusCode = StatusCode.BadRequest;
+    }
 
-        response.statusCode = statusCode;
-        if (responseError) {
-          return reject(responseError);
-        } else {
-          return resolve(response);
-        }
-      },
-    );
+    response.statusText = statusTextLookup[response.statusCode];
+    abstractConvertResponse(response, adapterArgs);
+    if (!isOK(response)) {
+      throw new ShopifyErrors.InvalidWebhookError({
+        message: errorMessage,
+        response,
+      });
+    }
```

Comment:
As per above, would this be. Also, I think we may need an `await` when converting?

```suggestion
    response.statusText = statusTextLookup[response.statusCode];
    const returnResponse = await abstractConvertResponse(response, adapterArgs);
    if (!isOK(response)) {
      throw new ShopifyErrors.InvalidWebhookError({
        message: errorMessage,
        response,
      });
    }
    
    return returnResponse;
```